[MORTY-2045] React Native Backpack Dialog

felixarpa opened the pull request:
<!--
Thanks for contributing to Backpack :pray:
Please include a description of the changes you are introducing and some screenshots if appropriate.
-->

# React Native Backpack Dialog

`BpkDialog` API for React Native.

Dialogs inform users about a specific task and may contain critical information, or require decisions or acknowledgment.

+ [x] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-react-native/blob/master/CONTRIBUTING.md)


_If you are curious about how we review, please read through the [code review guidelines](https://github.com/Skyscanner/backpack/blob/master/CODE_REVIEW_GUIDELINES.md)_

tiagoengel commented on packages/react-native-bpk-component-dialog/src/BpkDialog.js:
```diff
@@ -0,0 +1,68 @@
+const BpkDialog = (props: Props) => {
+  const { icon, ...rest } = props;
+
+  if (!icon.iconId.startsWith('bpk_')) {
```
I wouldn't do this like. I think is better to use the `icons` export from bpk icons and those names so it's more familiar with RN folks, and then adjust the name here. 

E.g.:
```
import { icons } from `react-native-bpk-component-icon`

icon={icons.beer}
```

And here you just rename it to `bpk_${iconId.replace(/-/g, '_')}`

felixarpa replied:
Should I do the same with the color?

felixarpa commented on packages/react-native-bpk-component-dialog/src/android/build.gradle:
```diff
@@ -0,0 +1,56 @@
+      // see: https://docs.gradle.org/current/userguide/declaring_dependencies.html#sub:declaring_dependency_rich_version
+      // see: https://docs.oracle.com/middleware/1212/core/MAVEN/maven_version.htm#MAVEN8855
+      // see: http://ant.apache.org/ivy/history/2.1.0/settings/version-matchers.html
+      strictly "[7.4,9.0)"
```
The whole file?

tiagoengel replied:
no just the `strictly` config. You can replace it with `implementation("com.github.skyscanner:backpack-android:8.0.0")` for example

georgegillams commented on packages/react-native-bpk-component-dialog/src/NativeDialog.ios.js:
```diff
@@ -0,0 +1,120 @@
+}
+
+BpkDialog.propTypes = commonPropTypes;
+// BpkDialog.defaultProps = commonDefaultProps;
```
Can be removed 👍

georgegillams commented on packages/react-native-bpk-component-dialog/src/NativeDialog.ios.js:
```diff
@@ -0,0 +1,120 @@
+  componentDidMount() {
+    if (BpkDialogEventEmitter) {
+      const { actions, scrimAction } = this.props;
+      this.eventSubscription = BpkDialogEventEmitter.addListener(
```
I'd be interested to know the rationale behind using event emitters, instead of allowing regular RN Button elements to be passed to the component. I'm sure there's a good reason.

felixarpa replied:
I used the event emitters because is how RN does it with the Modal:
https://github.com/facebook/react-native/blob/9bb5c32de9ef06942e38ea6cf5d9000c37741c37/Libraries/Modal/Modal.js#L180-L191

felixarpa commented on packages/react-native-bpk-component-dialog/src/ios/DialogBridge/RCTBPKDialog.h:
```diff
@@ -0,0 +1,59 @@
+
+#import <React/RCTInvalidating.h>
+#import <React/RCTView.h>
+#import <Backpack/Dialog.h>
```
I was not able to make this imports like `@class ...`, so I just removed those that are not used in both sides, the `.m` and the `.h` file.

felixarpa commented:
Hello @tiagoengel,

_It stops working when scrim is clicked_: I added the scrim action for android, so now it does not stop working. It will stop working if the developer does not update the state in the scrim action. That happens when the status still `isOpen: true` after the 
_No way to add an action that doesn’t hide the dialog_: That was because the Native part was calling `hide()` before invoking the callback. That will not happen anymore.
Right now in `stories.js` I added a line to every callback to set state `isOpen: false` right after calling the action's callback. If you remove this, the dialog will not hide after each alert.
_No way to add icons to buttons. This can wait if is not a requirement for your usage_: I'm going to skip this because it's not required now.

tiagoengel commented on android/gradle/wrapper/gradle-wrapper.properties:
```diff
@@ -1,6 +0,0 @@
-#Thu Mar 14 09:13:15 GMT 2019
```
This should not be removed

felixarpa commented on packages/react-native-bpk-component-dialog/src/android/src/main/java/net/skyscanner/backpack/reactnative/dialog/RNDialog.kt:
```diff
@@ -0,0 +1,102 @@
+  fun setUpActions() {
+    if (dirty) {
+      dialog.setOnDismissListener {
+        onAction?.invoke(SCRIM_CLOSED)
```
I didn't know that method existed 😮

felixarpa commented:
@tiagoengel thanks for your review 😄
I fixed everything you commented and implemented the `setCancelOnTocuchOutside()`. I also added a story for that case to check that it does not dismiss.

k0nserv commented on packages/react-native-bpk-component-dialog/src/BpkDialog.js:
```diff
@@ -0,0 +1,59 @@
+    const index = event.nativeEvent.actionIndex;
+    if (index >= 0 && index < actions.length) {
+      actions[index].callback();
+    } else if (index === SCRIM_CLOSED) {
```
Is this just how Android works actually? i.e. you only get the one way to communicate with native?

felixarpa replied:
Yes, this is only for android, I can create another action like `topTouchCancel`
What do you mean with `the other way`?
I also think could be good to move it to `Native.android.js`

felixarpa commented on ios/native.xcodeproj/project.pbxproj:
```diff
@@ -319,8 +320,10 @@
+				English,
 				en,
 				Base,
+				"he-IL",
```
I think it was added automatically

felixarpa replied:
Yes, Xcode adds that automatically, I'll remove it before pushing

felixarpa commented on packages/react-native-bpk-component-dialog/src/NativeDialog.android.js:
```diff
@@ -0,0 +1,58 @@
+          .toUpperCase()}${icon.iconColor.slice(1)}`,
+      }}
+      scrimEnabled={scrimAction.enabled}
+      {...rest}
```
The native event emitter is the `onChange` in `BpkDialog.js`

felixarpa commented on packages/react-native-bpk-component-dialog/src/ios/DialogBridge/RCTBPKDialogManager.m:
```diff
@@ -0,0 +1,110 @@
+
+    bpkDialog.dialogController.scrimAction = scrimAction;
+
+    for (NSUInteger i = 0; i < bpkDialog.actions.count; ++i) {
```
The index of the action is sent to the RN side to identify the action to trigger.

k0nserv commented:
Closing and merging via command line instead.
